File: src/protocol/commands/gif.ts
/**
 * GIF upload payload builder.
 *
 * GIF data is split into 4096-byte chunks. Each chunk gets a 16-byte header:
 *   [0-1]:  u16 LE = header_len + chunk_data_len
 *   [2]:    0x01 (static)
 *   [3]:    0x00 (static)
 *   [4]:    flag (0 for first chunk, 2 for subsequent)
 *   [5-8]:  u32 LE = total GIF data length
 *   [9-12]: u32 LE = CRC32 of entire GIF data
 *   [13-15]: 0x05, 0x00, 0x0D (static footer)
 *
 * Each chunk is sent separately with BLE write-with-response.
 */

import { GIF_CHUNK_SIZE, GIF_HEADER_SIZE } from "../constants.ts";
import { crc32 } from "../../utils/crc32.ts";

export function createGifPayloads(gifData: Uint8Array): Uint8Array[] {
	const crc = crc32(gifData);
	const chunks = splitIntoChunks(gifData, GIF_CHUNK_SIZE);
	const payloads: Uint8Array[] = [];

	for (let i = 0; i < chunks.length; i++) {
		const chunk = chunks[i]!;
		const header = new Uint8Array(GIF_HEADER_SIZE);
		const view = new DataView(header.buffer);

		const chunkLen = GIF_HEADER_SIZE + chunk.length;
		view.setUint16(0, chunkLen, true);
		header[2] = 0x01;
		header[3] = 0x00;
		header[4] = i > 0 ? 2 : 0;
		view.setUint32(5, gifData.length, true);
		view.setUint32(9, crc, true);
		header[13] = 0x05;
		header[14] = 0x00;
		header[15] = 0x0d;

		const payload = new Uint8Array(chunkLen);
		payload.set(header, 0);
		payload.set(chunk, GIF_HEADER_SIZE);
		payloads.push(payload);
	}

	return payloads;
}

function splitIntoChunks(data: Uint8Array, chunkSize: number): Uint8Array[] {
	const chunks: Uint8Array[] = [];
	for (let i = 0; i < data.length; i += chunkSize) {
		chunks.push(data.slice(i, i + chunkSize));
	}
	return chunks;
}

File: rust-cli/src/protocol/commands/gif.rs
use crate::protocol::constants::{GIF_CHUNK_SIZE, GIF_HEADER_SIZE};

pub fn create_gif_payloads(gif_data: &[u8]) -> Vec<Vec<u8>> {
    let crc = crc32fast::hash(gif_data);
    let chunks: Vec<&[u8]> = gif_data.chunks(GIF_CHUNK_SIZE).collect();
    let mut payloads = Vec::new();

    for (i, chunk) in chunks.iter().enumerate() {
        let chunk_len = GIF_HEADER_SIZE + chunk.len();
        let mut payload = vec![0u8; chunk_len];

        // Header
        payload[0..2].copy_from_slice(&(chunk_len as u16).to_le_bytes());
        payload[2] = 0x01;
        payload[3] = 0x00;
        payload[4] = if i > 0 { 0x02 } else { 0x00 };
        payload[5..9].copy_from_slice(&(gif_data.len() as u32).to_le_bytes());
        payload[9..13].copy_from_slice(&crc.to_le_bytes());
        payload[13] = 0x05;
        payload[14] = 0x00;
        payload[15] = 0x0D;

        // Data
        payload[GIF_HEADER_SIZE..].copy_from_slice(chunk);
        payloads.push(payload);
    }

    payloads
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_gif_payloads_single_chunk() {
        let data = vec![0xCC; 200];
        let payloads = create_gif_payloads(&data);
        assert_eq!(payloads.len(), 1);
        assert_eq!(payloads[0].len(), 216);
        assert_eq!(payloads[0][4], 0x00);
    }

    #[test]
    fn test_create_gif_payloads_multi_chunk() {
        let data = vec![0xDD; 8200];
        let payloads = create_gif_payloads(&data);
        assert_eq!(payloads.len(), 3);
        assert_eq!(payloads[0][4], 0x00);
        assert_eq!(payloads[1][4], 0x02);
        assert_eq!(payloads[2][4], 0x02);
    }
}
